components/builder-api/src/server/resources/origins.rs

```diff
@@ -269,15 +272,108 @@ fn delete_origin(req: HttpRequest, path: Path<String>, state: Data<AppState>) ->
         Err(err) => return err.into(),
     };
 
-    match Origin::delete(&origin, &*conn).map_err(Error::DieselError) {
-        Ok(_) => HttpResponse::NoContent().into(),
+    // Prior to passing the deletion request to the backend, we validate
+    // that the user has already cleaned up the most critical origin data.
+    match origin_delete_preflight(&origin, &*conn) {
+        Ok(_) => {
+            match Origin::delete(&origin, &*conn).map_err(Error::DieselError) {
+                Ok(_) => HttpResponse::NoContent().into(),
+                Err(err) => {
+                    debug!("Origin {} deletion failed! err = {}", origin, err);
+                    // We do not want to expose any database details from diesel
+                    // thus we simply return a 409 with an empty body.
+                    HttpResponse::new(StatusCode::CONFLICT)
+                }
+            }
+        }
         Err(err) => {
-            debug!("Origin {} is not deletable, err = {}", origin, err);
-            HttpResponse::new(StatusCode::CONFLICT)
+            debug!("Origin preflight determined that {} is not deletable, err = {}!",
+                   origin, err);
+            // Here we want to enrich the http response with a sanitized error
+            // by returning a 409 with a helpful message in the body.
+            HttpResponse::with_body(StatusCode::CONFLICT, Body::from_message(format!("{}", err)))
         }
     }
 }
 
+#[allow(clippy::needless_pass_by_value)]
+fn origin_delete_preflight(origin: &str, conn: &PgConnection) -> Result<()> {
+    match Project::count_origin_projects(&origin, &*conn) {
+        Ok(0) => {}
+        Ok(count) => {
+            let err = format!("There are {} projects remaining in origin {}. Must be zero.",
+                              count, origin);
+            return Err(Error::BuilderCore(OriginDeleteError(err)));
+        }
+        Err(e) => return Err(Error::DieselError(e)),
+    };
+
+    match OriginMember::count_origin_members(&origin, &*conn) {
+        // allow 1 - the origin owner
```

Comment:
Question: Could there ever a case where there is a single member that _isn't_ the owner?  and would that matter here? 

Thinking through this as I type, I suspect the answer is no, as you need to be an owner to delete an origin. It could be a problem if https://github.com/habitat-sh/builder/blob/021a5216584b194f6b6e89d0b4f73487dd2fd74d/components/builder-api/src/server/resources/origins.rs#L264 ever got moved after the pre-flight, or had a logic error in it.

Reply:
I think not - we have protection for all of the origin transfer and various behaviors to protect from this ever happening. An orphaned origin would be a bug.

Reply:
I concur. If we ever changed the behavior, the first feedback would be from API tests that begin failing in the developer's studio.